### PR TITLE
Rework problem and correction messages

### PR DESCRIPTION
**Pull request by bwilkerson**
This is a small sample of the changes to messages that I'm proposing to make. Feedback is absolutely welcome on both the messages themselves and on the process (for example, if it would be easier to put all the messages in a doc than in a PR).

### CODE REVIEW

**Review by pq — COMMENTED** [review]
In general I really like where this is going!

Perhaps we should create a proposal and loop some more folks into a broader conversation?

File: lib/src/rules/avoid_empty_else.dart [hunk 1]
```diff
@@ -24,6 +24,14 @@ else ;
 ''';
 
 class AvoidEmptyElse extends LintRule {
+  static const LintCode code = LintCode('avoid_empty_else',
+      "Empty statements are not allowed in an 'else' clause.",
+      correctionMessage:
+          'Try removing the empty statement or removing the else clause.');
+
+  @override
+  LintCode get lintCode => AvoidEmptyElse.code;
```
[the remark this answers is not in the document]
Comment (bwilkerson) [reply lib/src/rules/avoid_empty_else.dart:33]: (I'm guessing by "super call" you mean in the constructor.) Passing it in to the constructor would only work in those cases where there's a single lint code, so I don't know that passing it in is a general enough mechanism. Also, I question the value of storing the code in an instance field of every object when it's already being stored in a static field. Seems inefficient.

File: lib/src/rules/avoid_relative_lib_imports.dart [hunk 2]
```diff
@@ -78,7 +87,7 @@ class _Visitor extends SimpleAstVisitor<void> {
   @override
   void visitImportDirective(ImportDirective node) {
     if (isRelativeLibImport(node)) {
-      rule.reportLint(node.uri);
+      rule.reportLint(node.uri, errorCode: AvoidRelativeLibImports.code);
```
[the remark this answers is not in the document]
Comment (bwilkerson) [reply lib/src/rules/avoid_relative_lib_imports.dart:90]: I hadn't realized that, thanks. I'll back out those changes.

File: lib/src/rules/hash_and_equals.dart [hunk 3]
```diff
@@ -95,14 +95,16 @@ class _Visitor extends SimpleAstVisitor<void> {
...
     }
     if (hash != null && eq == null) {
       if (hash is MethodDeclaration) {
-        rule.reportLint(hash.name, errorCode: hashMemberCode);
+        rule.reportLint(hash.name,
+            errorCode: HashAndEquals.code, arguments: ['==', 'hashCode']);
       } else if (hash is FieldDeclaration) {
         rule.reportLint(getFieldIdentifier(hash, 'hashCode'),
-            errorCode: hashMemberCode);
+            errorCode: HashAndEquals.code, arguments: ['==', 'hashCode']);
```
[the remark this answers is not in the document]
Comment (bwilkerson) [reply lib/src/rules/hash_and_equals.dart:107]: Reverse what? (I like that I can select less than a whole line in Gerrit.)

**Comment by bwilkerson** [comment]
> Perhaps we should create a proposal and loop some more folks into a broader conversation?

I'm not sure I followed that. Proposal for what? And who do you think we should loop in?

File: lib/src/rules/hash_and_equals.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/hash_and_equals.dart:107]: should `['==', 'hashCode']` be `['hashCode', '==']` ?

File: lib/src/rules/avoid_empty_else.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/avoid_empty_else.dart:33]: Right.  That's what I meant.   It's true that it'd only work for the single lint code case but that's the (vast?) majority of lints at the moment.  I'm neutral about whether it's better.   I was just pointing out an opportunity to align with or break from an existing pattern.

File: lib/src/rules/hash_and_equals.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/hash_and_equals.dart:107]: Sorry!  It's correct.  I was reading the diff wrong and thought the order was the same for both cases.  Sorry for the noise!

**Comment by bwilkerson** [comment]
PTAL

**Review by pq — APPROVED** [review]
👍

**Pull request merged**